lib/internal/fetch/headers.js

```
@@ -0,0 +1,291 @@
+    value,
+    ''
+  );
+  validateHeaderValue(name, normalizedHeaderValue);
+  return normalizedHeaderValue;
+}
+
+function isHeaders(object) {
+  return kHeadersList in object;
+}
+
+function fill(headers, object) {
+  if (isHeaders(object)) {
+    // Object is instance of Headers
+    headers[kHeadersList] = ArrayPrototypeSlice(object[kHeadersList]);
+  } else if (ArrayIsArray(object)) {
+    // Support both 1D and 2D arrays of header entries
+    if (ArrayIsArray(object[0])) {
+      // Array of arrays
+      for (let i = 0; i < object.length; i++) {
+        if (object[i].length !== 2) {
+          throw new ERR_INVALID_ARG_VALUE('init', object[i],
+                                          'is not of length 2');
+        }
+        headers.append(object[i][0], object[i][1]);
+      }
+    } else if (typeof object[0] === 'string' || Buffer.isBuffer(object[0])) {
+      // Flat array of strings or Buffers
+      if (object.length % 2 !== 0) {
+        throw new ERR_INVALID_ARG_VALUE('init', object,
+                                        'is not even in length');
+      }
+      for (let i = 0; i < object.length; i += 2) {
+        headers.append(
+          object[i].toString('utf-8'),
+          object[i + 1].toString('utf-8')
+        );
+      }
+    } else {
+      // All other array based entries
+      throw new ERR_INVALID_ARG_VALUE(
+        'init',
+        object,
+        'is not a valid array entry'
+      );
+    }
+  } else if (!isBoxedPrimitive(object)) {
+    // Object of key/value entries
+    const entries = ObjectEntries(object);
+    for (let i = 0; i < entries.length; i++) {
+      headers.append(entries[i][0], entries[i][1]);
+    }
+  }
+}
+
+class Headers {
+  constructor(init = {}) {
+    // Fail silently on primitives
+    const typeofInit = typeof init;
+    if (typeofInit === 'number' ||
+     typeofInit === 'string' ||
+     typeofInit === 'boolean' ||
+     typeofInit === 'function') return;
```

Why fail silently? Browsers seem to be pretty vocal about it:

```js
new Headers(1); // TypeError
new Headers(1n); // TypeError
new Headers(''); // TypeError
new Headers(true); // TypeError
new Headers(Symbol()); // TypeError

new Headers(() => {}); // no type error
```

---

Ahh I must've misunderstood the spec